File: src/MinStack.java
import java.util.ArrayList;
import java.util.List;

public class MinStack {

    private final List<Integer> numbers = new ArrayList<>();

    /** initialize your data structure here. */
    public MinStack() {
    }

    public void push(int x) {
        numbers.add(x);
    }

    public void pop() {
        numbers.remove(numbers.size() - 1);
    }

    public int top() {
        return numbers.get(numbers.size() - 1);
    }

    public int getMin() {
        int min = Integer.MAX_VALUE;
        for (int element : numbers) {
            min = Math.min(min, element);
        }
        return min;
    }
}

File: python/min_stack.py
class Node:
    def __init__(self, val, min, previous):
        self.val = val
        self.next = next
        self.min = min
        self.previous = previous

class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.current = None

    def push(self, val: int) -> None:
        if self.current is None:
            self.current = Node(val, val, None)
        else:
            self.current.next = Node(val, min(val, self.current.min), self.current)
            self.current = self.current.next

    def pop(self) -> None:
        val = self.current.val
        self.current = self.current.previous
        return val

    def top(self) -> int:
        return self.current.val

    def getMin(self) -> int:
        return self.current.min
